### ISSUE TITLE
Improve RepartitionExec for better query performance

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#7001]

**Comment by scsmithr** [comment]
👍 for this.

I poked around with removing the gate in the distribution channel implementation. My challenges with the gate implementation was high lock contention, as well as high memory usage when inputs are highly skewed (e.g. one channel received the bulk of the batches, but the gate still let batches in since other channels were empty).

Some experimentation here: https://github.com/GlareDB/arrow-datafusion/blob/repart-perf/datafusion/core/src/physical_plan/repartition/distributor_channels.rs. Note that this hangs on some of the repartition tests, so there's likely a bug in the futures logic.

With a high enough channel buffer size, we start to see similar performance increases seen in the flume pr:

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ repart-perf ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 309.43ms │    309.24ms │     no change │
│ QQuery 2     │  47.65ms │     48.91ms │     no change │
│ QQuery 3     │ 129.55ms │    109.39ms │ +1.18x faster │
│ QQuery 4     │  81.07ms │     51.09ms │ +1.59x faster │
│ QQuery 5     │ 158.61ms │    123.25ms │ +1.29x faster │
│ QQuery 6     │  79.99ms │     80.40ms │     no change │
│ QQuery 7     │ 218.47ms │    202.31ms │ +1.08x faster │
│ QQuery 8     │ 184.88ms │    170.81ms │ +1.08x faster │
│ QQuery 9     │ 266.72ms │    212.60ms │ +1.25x faster │
│ QQuery 10    │ 207.46ms │    149.29ms │ +1.39x faster │
│ QQuery 11    │  44.44ms │     44.40ms │     no change │
│ QQuery 12    │ 150.27ms │    119.97ms │ +1.25x faster │
│ QQuery 13    │ 227.68ms │    229.70ms │     no change │
│ QQuery 14    │ 113.26ms │    111.97ms │     no change │
│ QQuery 15    │  81.23ms │     81.10ms │     no change │
│ QQuery 16    │  51.57ms │     51.50ms │     no change │
│ QQuery 17    │ 338.65ms │    305.00ms │ +1.11x faster │
│ QQuery 18    │ 379.12ms │    370.23ms │     no change │
│ QQuery 19    │ 233.23ms │    234.30ms │     no change │
│ QQuery 20    │ 132.11ms │    129.32ms │     no change │
│ QQuery 21    │ 324.30ms │    256.37ms │ +1.26x faster │
│ QQuery 22    │  50.06ms │     43.18ms │ +1.16x faster │
└──────────────┴──────────┴─────────────┴───────────────┘
```

(On an M2 Macbook Air)

Not trying to push for any one approach, and this was mostly just an experiment.

**Comment by alamb** [comment]
I have been dreaming about this -- I may try and bash out a prototype in the coming weeks 

I find it a very interesting intellectual challenge "can we make a tokio based scheduler build a system that take into account NUMA and cache locality" -- I think we can and this would be an excellent opportunity to show it

**Comment by Dandandan** [comment]
That would be amazing @alamb very interested o see what you come up with :)

**Comment by alamb** [comment]
Also @wolffcm noted elsewhere that

> I see a lot of DF plans that contain something like this:
```
    RepartitionExec: partitioning=Hash(...), input_partitions=12
      RepartitionExec: partitioning=RoundRobinBatch(12), input_partitions=2
```

> At first I thought this was an unintentional artifact of how physical planning happens, but now it seems to me that the intention of the RoundRobin repartitioning is to ensure we paralellize the Hash repartitioning, which can be expensive because it looks at each row. Do I have that right?

I believe this is correct, and perhaps would be something to improve in any improvement

**Comment by wolffcm** [comment]
To add a little bit more context to the above, the `RoundRobin` repartitioning is added as part of the `Repartition` optimizer pass, and then the `Hash` repartitioning is added by `EnforceDistribution` (because a subsequent operator requires it)

**Comment by Dandandan** [comment]
The current plan to do round robin before hashing seems optimal to me, like @wolffcm I think also suggests? The other way around would reduce the parallelism of hashing.

**Comment by alamb** [comment]
I was thinking of a unified roundrobin / hashing repartitioner that was in terms of inputs / outputs. I may have mislead myself in my mind and will have to see how the code plays out

**Comment by alamb** [comment]
Some other notes from a DM I am bringing here for visibility:

 DataFusions pattern can be described as a "classic pull scheduler with volcano style exchange operators":
* pull scheduler == Tokio
* and exchange operators == Repartition

My theory (which I would say not everyone agrees with yet) is that you can get all the claimed benefits of "morsel driven parallelism" (aka a push scheduler) from a "classic pull scheduler with exchange operators"


The **claimed** benefits of morsel driven parallelism of the classic approach are:
1. Better NUMA / cache locality
2. Near linear scaling (so if you double core count to 32 to 64 expect to see a linear 2x speed up)
3. Better runtime optimization by scaling up/down cores, but I think this is less important in the real world (as you most often run out of memory long before you run out of CPU)

Basically I think it is totally reasonable for DataFusion to scale linearly to high core counts -- like 128 / 256. I just need to prove that is the case and I suspect it will take some finagling with the current Repartitioning code

**Comment by scsmithr** [comment]
With the current architecture, I think the easiest thing to do here is probably to not try to have a one-to-one mapping of task to core, and lean into tokio for scheduling to try to get good resource utilization.

I did a quick experiment for this by increasing the number of partitions to use in the repartition optimization rule. Currently it gets set to `target_partitions`, and I just multiplied it by 3.

The diff:

```diff
diff --git a/datafusion/core/src/physical_optimizer/repartition.rs b/datafusion/core/src/physical_optimizer/repartition.rs
index aa48fd77a..b4ef8c3a2 100644
--- a/datafusion/core/src/physical_optimizer/repartition.rs
+++ b/datafusion/core/src/physical_optimizer/repartition.rs
@@ -288,7 +288,7 @@ impl PhysicalOptimizerRule for Repartition {
         plan: Arc<dyn ExecutionPlan>,
         config: &ConfigOptions,
     ) -> Result<Arc<dyn ExecutionPlan>> {
-        let target_partitions = config.execution.target_partitions;
+        let target_partitions = config.execution.target_partitions * 3;
         let enabled = config.optimizer.enable_round_robin_repartition;
         let repartition_file_scans = config.optimizer.repartition_file_scans;
         let repartition_file_min_size = config.optimizer.repartition_file_min_size;
```

And some benchmarks using TPCH SF=10

Macbook air m2:
```
Comparing main and three
--------------------
Benchmark tpch.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃     three ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 3615.39ms │ 1745.94ms │ +2.07x faster │
│ QQuery 2     │  447.68ms │  450.67ms │     no change │
│ QQuery 3     │ 1320.39ms │  730.22ms │ +1.81x faster │
│ QQuery 4     │  726.73ms │  394.54ms │ +1.84x faster │
│ QQuery 5     │ 1619.80ms │ 1064.71ms │ +1.52x faster │
│ QQuery 6     │  741.31ms │  371.64ms │ +1.99x faster │
│ QQuery 7     │ 3269.57ms │ 2528.63ms │ +1.29x faster │
│ QQuery 8     │ 1972.34ms │ 1200.81ms │ +1.64x faster │
│ QQuery 9     │ 3032.73ms │ 2252.82ms │ +1.35x faster │
│ QQuery 10    │ 2045.88ms │ 1381.35ms │ +1.48x faster │
│ QQuery 11    │  466.09ms │  484.14ms │     no change │
│ QQuery 12    │ 1400.27ms │  652.58ms │ +2.15x faster │
│ QQuery 13    │ 2088.00ms │ 1181.92ms │ +1.77x faster │
│ QQuery 14    │ 1047.22ms │  551.92ms │ +1.90x faster │
│ QQuery 15    │  802.09ms │  650.86ms │ +1.23x faster │
│ QQuery 16    │  429.19ms │  418.61ms │     no change │
│ QQuery 17    │ 4392.38ms │ 4516.79ms │     no change │
│ QQuery 18    │ 7849.27ms │ 7245.33ms │ +1.08x faster │
│ QQuery 19    │ 2180.33ms │ 1012.43ms │ +2.15x faster │
│ QQuery 20    │ 1657.01ms │ 1110.12ms │ +1.49x faster │
│ QQuery 21    │ 3867.34ms │ 3671.96ms │ +1.05x faster │
│ QQuery 22    │  490.25ms │  443.54ms │ +1.11x faster │
└──────────────┴───────────┴───────────┴───────────────┘
```

GCP n1 vm (8 cores/16 threads, 60G memory):
```
Comparing main and three
--------------------
Benchmark tpch.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃       main ┃      three ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  8947.17ms │  3274.31ms │ +2.73x faster │
│ QQuery 2     │  1562.68ms │  1519.65ms │     no change │
│ QQuery 3     │  3839.36ms │  1758.32ms │ +2.18x faster │
│ QQuery 4     │  1995.52ms │   967.04ms │ +2.06x faster │
│ QQuery 5     │  4906.24ms │  2933.39ms │ +1.67x faster │
│ QQuery 6     │  1941.19ms │   710.94ms │ +2.73x faster │
│ QQuery 7     │  9391.94ms │  6423.27ms │ +1.46x faster │
│ QQuery 8     │  5503.67ms │  2735.65ms │ +2.01x faster │
│ QQuery 9     │  8840.15ms │  6129.22ms │ +1.44x faster │
│ QQuery 10    │  5807.23ms │  3804.27ms │ +1.53x faster │
│ QQuery 11    │  1699.21ms │  1744.50ms │     no change │
│ QQuery 12    │  2993.60ms │  1277.92ms │ +2.34x faster │
│ QQuery 13    │  4907.55ms │  2585.08ms │ +1.90x faster │
│ QQuery 14    │  2633.90ms │  1114.95ms │ +2.36x faster │
│ QQuery 15    │  2085.41ms │  1176.97ms │ +1.77x faster │
│ QQuery 16    │  1683.76ms │  1572.89ms │ +1.07x faster │
│ QQuery 17    │  7322.44ms │  6412.73ms │ +1.14x faster │
│ QQuery 18    │ 15950.64ms │ 16131.98ms │     no change │
│ QQuery 19    │  4936.98ms │  1834.62ms │ +2.69x faster │
│ QQuery 20    │  5126.77ms │  3039.32ms │ +1.69x faster │
│ QQuery 21    │ 14114.41ms │  9196.05ms │ +1.53x faster │
│ QQuery 22    │  1449.16ms │  1206.74ms │ +1.20x faster │
└──────────────┴────────────┴────────────┴───────────────┘
```

**Comment by alamb** [comment]
@scsmithr  this is great data -- thank you  -- and it does imply there is substantial headroom for performance improvement by using more cores. t is interesting that some queries go 2x as fast -- it is almost like the default value for target_partitions was 1/2 the actual available cores

However, the config seems to be doing the right thing:

https://github.com/apache/arrow-datafusion/blob/ba51b6a334de813a2b90998b72f14281460e8f06/datafusion/common/src/config.rs#L218

🤔

**Comment by ozankabak** [comment]
That's very interesting, do you have an idea what may be going on @alamb?

**Comment by scsmithr** [comment]
Much of the improvement here is from repartitioning in the ParquetExec. Since that's io bound, having more tasks than cores improves resource utilization if that's what the rest of query is waiting on. It might be reasonable to introduce a `io_target_partitions` config variable specifically for repartitioning at the io layer.

**Comment by ozankabak** [comment]
Makes sense 👍

**Comment by alamb** [comment]
@scsmithr  how did you get the numbers you report in https://github.com/apache/arrow-datafusion/issues/7001#issuecomment-1675625184 ?

I believe the benchmarks default to using only 2 cores: https://github.com/apache/arrow-datafusion/blob/ed85abbb878ef3d60e43797376cb9a40955cd89a/benchmarks/src/options.rs#L29-L31

So if you ran the tests without specifying `--partitions 16` or whatever they probably aren't using eveything

I wonder if we should increase the default number of partitions to the core count in the benchmarks 🤔

**Comment by scsmithr** [comment]
> So if you ran the tests without specifying --partitions 16 or whatever they probably are only using 2 cores

Yep completely missed that option. I was running `./benchmarks/bench.sh run tpch10` without setting this flag. Setting `--partitions 8` on my macbook negates any improvement.

I'm coming at this with trying to speed up reading unpartitioned parquet files from GCS, and there's improvements when setting target partitions beyond number of cores. The above benchmark numbers were me trying to see if that applied generally, and not just in this specific case. Turns out, it doesn't apply generally... still investigating why I'm seeing speed ups with GCS.

> I wonder if we should increase the default number of partitions to the core count in the benchmarks 🤔

Would making `partitions` an Option make sense? If unset, default to whatever is the default in the session config. I had incorrectly assumed that this is what would happen when running `bench.sh`.

**Comment by alamb** [comment]
> Would making partitions an Option make sense? If unset, default to whatever is the default in the session config. I had incorrectly assumed that this is what would happen when running bench.sh.

I think it would make sense as it better reflects the default configuration DataFusion is run in. I wonder if the benchmarks originally were trying to isolate the specific algorithms and avoid the additional noise that multi-cores brought. However, given that they run with 2 cores now it seems they get the worst of both worlds

@scsmithr  do you have time to make a PR with that change?

**Comment by Dandandan** [comment]
I'm currently running some experiments with this - hope to have some results later today.

**Comment by Dandandan** [comment]
Doing experiments in https://github.com/apache/datafusion/pull/13699 if anyone is interested, results seem to look very promising (need to double check).

**Comment by Dandandan** [comment]
This is for TPC-H SF=1:
```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ repartition_exec_poc ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 128.14ms │             104.67ms │ +1.22x faster │
│ QQuery 2     │  55.15ms │              37.12ms │ +1.49x faster │
│ QQuery 3     │  57.28ms │              56.86ms │     no change │
│ QQuery 4     │  39.74ms │              37.97ms │     no change │
│ QQuery 5     │  77.21ms │              75.23ms │     no change │
│ QQuery 6     │  21.25ms │              22.48ms │  1.06x slower │
│ QQuery 7     │  93.89ms │              88.99ms │ +1.06x faster │
│ QQuery 8     │  80.49ms │              71.86ms │ +1.12x faster │
│ QQuery 9     │ 115.25ms │             101.68ms │ +1.13x faster │
│ QQuery 10    │  97.48ms │             106.67ms │  1.09x slower │
│ QQuery 11    │  40.27ms │              25.46ms │ +1.58x faster │
│ QQuery 12    │  54.96ms │              48.50ms │ +1.13x faster │
│ QQuery 13    │ 121.18ms │              82.66ms │ +1.47x faster │
│ QQuery 14    │  37.72ms │              40.23ms │  1.07x slower │
│ QQuery 15    │  49.46ms │              48.15ms │     no change │
│ QQuery 16    │  36.71ms │              24.40ms │ +1.50x faster │
│ QQuery 17    │ 111.15ms │             100.77ms │ +1.10x faster │
│ QQuery 18    │ 167.05ms │             152.32ms │ +1.10x faster │
│ QQuery 19    │  66.07ms │              59.23ms │ +1.12x faster │
│ QQuery 20    │  58.43ms │              59.72ms │     no change │
│ QQuery 21    │ 133.90ms │             119.76ms │ +1.12x faster │
│ QQuery 22    │  36.96ms │              28.42ms │ +1.30x faster │
└──────────────┴──────────┴──────────────────────┴───────────────┘
```

**Comment by Dandandan** [comment]
It seems the improvement comes not from the adaptivity but from improved balancing, so I'm probably opening a simplified PR with similar improvement.

**Comment by pranavJibhakate** [comment]
@alamb  @ozankabak What are the ways to make RoundRobin more NUMA aware.
I could come up with only this approach
I was reading about how we could pin threads to CPUs so we could add core id to the tasks details and schedule it only on those tasks and if the local task queue is empty it would work steal like tokio usually does?

**Comment by alamb** [comment]
> [@alamb](https://github.com/alamb) [@ozankabak](https://github.com/ozankabak) What are the ways to make RoundRobin more NUMA aware. I could come up with only this approach I was reading about how we could pin threads to CPUs so we could add core id to the tasks details and schedule it only on those tasks and if the local task queue is empty it would work steal like tokio usually does?

I don't think I have an answer for this one

We would probably have to try some approaches to see if anything works

**Comment by Rich-T-kid** [comment]
take

**Comment by Dandandan** [comment]
Do we have any ideas to improve this further? 
The coalescing has been improved recently, also the morsel scanning helps load balancing at the scan level (while still having some back pressure).
Also I added some improvement (strength reduce) and we improved hashing performance which all helped.

I think some other features like benefiting more from (almost) sorted data / clustering might be interesting here.

**Comment by Rich-T-kid** [comment]
> Do we have any ideas to improve this further?

no, not off the top of my head.

> I think some other features like benefiting more from (almost) sorted data / clustering might be interesting here.

i'll investigate and keep this in mind 👍 

related : https://github.com/apache/datafusion/issues/20773

**Comment by Rich-T-kid** [comment]
Okay I think I have a few ideas for how we can speed up RepartionExec. first lets do a review so everyone is on the same page.

<img width="2928" height="1362" alt="Image" src="https://github.com/user-attachments/assets/a0807852-4857-469d-8240-c39bc24c0b8e" />

At a high level this is how repartionExec, specifically the HashRepartion path.

1. hash the key columns. `M`,`N`,`K` in `Hash(M,N,K)` 
2.  Call `take_arrays()` once per input batch to produce one reordered batch, then slice() it into per-partition sub-batches (https://github.com/apache/datafusion/pull/22159)
3.  Each sub-batch is pushed into its output partition's `SharedCoalescer`, which buffers rows behind a Mutex until `target_batch_size` is reached
4.  When `target_batch_size` is reached, `push_and_drain()` returns the completed batch, which is then sent over the partition's async channel to the consumer side


The most CPU-intensive parts of this are `take_arrays()` and `push_batch()`, since every row passes through both. Key hashing only scales with the number of columns being hashed, and as @gabotechs found [here](https://github.com/apache/datafusion/pull/23720#issuecomment-5104141874), improvements there would only be marginal.


I've been working on a [series of take optimizations](https://github.com/apache/arrow-rs/issues/8879) in arrow-rs that build on top of one another nicely — [optimize take bool & take utf8View](https://github.com/apache/arrow-rs/pull/10813), [optimize take list](https://github.com/apache/arrow-rs/pull/10812), [optimize fsl](https://github.com/apache/arrow-rs/pull/10441), etc. The most important piece is the final one, [#10945](https://github.com/apache/arrow-rs/pull/10945), which has shown up to a 25% speedup.

Simply swapping the existing methods for the new ones introduced in that PR should yield significant improvements here, given how heavily RepartitionExec relies on take.

That leaves SharedCoalescer, which is a thin wrapper around [BatchCoalescer](https://arrow.apache.org/rust/arrow/compute/struct.BatchCoalescer). I've made a few attempts to [optimize](https://github.com/apache/arrow-rs/issues/7761) this, with little to no success, but I think the best return on effort is [Optimize BatchCoalescer::push_batch_with_indices](https://github.com/apache/arrow-rs/issues/8957) (by @Dandandan).

The core issue is that we're constrained by having to copy repeatedly. `take()` has to copy every byte from input to output, and BatchCoalescer (except for primitives, utf8View, and fixed-size binary) has to allocate 2x memory and then copy the bytes from source to destination. Right now this copy happens twice; simply reducing that to a single copy could be a significant win.

<img width="2828" height="1396" alt="Image" src="https://github.com/user-attachments/assets/e5fa6c96-88bb-4afb-8e92-c3fca814cfff" />

What Im proposing is to avoid the `take_arrays()` call in place for passing a zero-copy slice to `SharedCoalescer` as well as the indices that would have went to `take_arrays()`. then SharedCoacler would use `push_batch_with_indices()`. the issue with using this today is that this method is just a wrapper for `take_arrays()` https://github.com/apache/arrow-rs/blob/e491ce6c6f343ff9d7a235376d0353d12b365330/arrow-select/src/coalesce.rs#L291.

The first step towards this is updating the backing trait `InprogressArrays` to take advantage of `push_batch_with_indices()` by writing the intermediate state directly to the output buffer. That is introduced in this PR https://github.com/apache/arrow-rs/pull/11174.


Besides that I think the other largest wins are trying to speed up the concat kernel as this is the generic path (non-primitves/View types).

**Comment by Rich-T-kid** [comment]
this idea is somewhat similar to https://github.com/apache/datafusion/issues/15382 but we arent skipping `take()` + `batchCoalecer` as the issue proposes. we would keep the same indices `take()` would accept but instead pass it along to batchCoalecer. we don't need to change anything else about how the plan executes, the changes would be purely scooped to the repartition operator
